experimental/incremental/task.go

```diff
@@ -35,51 +36,65 @@ type Task struct {
+	done bool
+
+	// Intrusive linked list node for cycle detection.
+	path path
 }
 
 // Error adds errors to the current query, which will be propagated to all
 // queries which depend on it.
 //
-// This will not cause the query to fail; see [Task.Fail] for that.
-func (t *Task) Error(errs ...error) {
-	t.result.Errors = append(t.result.Errors, errs...)
-}
-
-// Fail marks the current query as failed. This will cause all queries that
-// depend on this one to also fail.
-//
-// This function will not return: it will cause a goroutine which calls it to
-// exit. See [Resolve] for notes on what this entails.
-func (t *Task) Fail(errs ...error) {
-	t.Error(errs...)
-	t.result.Failed = true
-	close(t.result.done)
-	runtime.Goexit()
+// This will not cause the query to fail; instead, [Query].Execute should
+// return false for the ok value to signal failure.
+func (t *Task) NonFatal(errs ...error) {
+	t.checkDone()
+	t.result.NonFatal = append(t.result.NonFatal, errs...)
 }
 
 // Resolve executes a set of queries in parallel. Each query is run on its own
 // goroutine.
 //
-// If any of the queries fails, or if the [context.Context] passed to the
-// [Run] call that spawned the [Task] is cancelled, this function calls
-// [runtime.Goexit]. This is not something callers should be concerned about,
-// because every call to a Query execution function is done in its own
-// goroutine, and only occurs when the return value of [Query].Execute will be
-// discarded.
+// If the context passed [Executor] expires, this will return [context.Cause].
+// The caller MUST immediately propagate this error to ensure the whole query
+// graph exits quickly. Failure to do so can result in deadlock or other
+// concurrent mayhem.
+//
+// If a cycle is detected for a given query, the query will automatically fail
+// and produce an [ErrCycle] for its fatal error. If the call to [Query].Execute
+// returns an error, that will be placed into the fatal error value, instead.
+//
+// Callers should make sure to check each result's fatal error before using
+// its value.
 //
-// Errors for each result are only those that occurred as a direct result of
-// query execution, and *does not* contain that query's transitive errors. This
-// is unlike the behavior of [Run].
-func Resolve[T any](caller Task, queries ...Query[T]) []Result[T] {
-	results := make([]Result[T], len(queries))
+// Non-fatal errors for each result are only those that occurred as a direct
+// result of query execution, and *does not* contain that query's transitive
+// errors. This is unlike the behavior of [Run], which only collects errors at
+// the very end. This ensures that errors are not duplicated, something that
+// is not possible to do mid-query.
+//
+// Note: this function really wants to be a method of [Task], but it isn't
+// because it's generic.
+func Resolve[T any](caller Task, queries ...Query[T]) (results []Result[T], expired error) {
+	caller.checkDone()
```

Comment:
I somewhat disagree that this should be an error, but I don't think it deeply matters.

Reply:
Actually no it has to be a panic, otherwise e.g. callers need to check for an error in NonFatal too. I think a panic is better, because it should only ever happen if you do something nasty like escape a `Task` into a global.